qa-tester/src/AdminBar.php

```diff
@@ -0,0 +1,223 @@
+	 *
+	 * @param WP_Dependencies $dependencies_object Dependencies object (WP_Scripts or WP_Styles).
+	 * @param string[]        $dependency_handles  Handles of the (enqueued) dependencies.
+	 * @return string[] All dependencies of dependencies.
+	 */
+	private function get_all_dependencies( WP_Dependencies $dependencies_object, $dependency_handles ) {
+		$original_handles_to_do     = $dependencies_object->to_do;
+		$dependencies_object->to_do = [];
+		$dependencies_object->all_deps( $dependency_handles, true );
+		$all_dependencies           = $dependencies_object->to_do;
+		$dependencies_object->to_do = $original_handles_to_do;
+		return $all_dependencies;
+	}
+
+	/**
+	 * Enqueue the plugin assets.
+	 */
+	public function enqueue_plugin_assets() {
+		// Only active if the admin bar is showing.
+		if ( ! is_admin_bar_showing() ) {
+			return;
+		}
+
+		$asset_file   = Plugin::get_path( 'assets/js/admin-bar.asset.php' );
+		$asset        = require $asset_file;
+		$dependencies = $asset['dependencies'];
+		$version      = $asset['version'];
+
+		$dependencies[] = 'hoverIntent';
+
+		// Enqueue scripts.
+		wp_enqueue_script(
+			self::ASSET_HANDLE,
+			Plugin::get_asset_url( 'js/admin-bar.js' ),
+			array_merge( [ 'admin-bar' ], $dependencies ),
+			$version,
+			true
+		);
+
+		// Enqueue styling for light DOM.
+		wp_enqueue_style(
+			self::ASSET_HANDLE,
+			Plugin::get_asset_url( 'css/admin-bar-light-dom-compiled.css' ),
+			[ 'admin-bar' ],
+			$version
+		);
+		wp_styles()->add_data( 'amp-qa-tester-admin-bar-style', 'rtl', 'replace' );
+
+		// Pass URL for CSS file to be loaded in shadow DOM.
+		$css_url = Plugin::get_asset_url( 'css/admin-bar-compiled' . ( is_rtl() ? '-rtl' : '' ) . '.css' );
+		wp_add_inline_script(
+			self::ASSET_HANDLE,
+			sprintf( 'var ampQaTester = %s;', wp_json_encode( compact( 'css_url' ) ) ),
+			'before'
+		);
+
+		// Explicitly mark all scripts and styles (and their dependencies) as being part of AMP dev mode.
+		// This is necessary because not all dependencies (and their recursive dependencies) will have a dependency
+		// on the admin-bar, and thus the AMP plugin won't opt them in to dev mode automatically.
+		$is_amp_request = function_exists( 'amp_is_request' ) ? amp_is_request() : is_amp_endpoint();
+		if ( $is_amp_request ) {
+			$script_dev_mode_handles = $this->get_all_dependencies( wp_scripts(), $dependencies );
+			add_filter(
+				'script_loader_tag',
+				function ( $tag, $handle ) use ( $script_dev_mode_handles ) {
+					if ( in_array( $handle, $script_dev_mode_handles, true ) ) {
+						$tag = preg_replace( '/(?<=<script)/', ' data-ampdevmode ', $tag );
+					}
+					return $tag;
+				},
+				10,
+				2
+			);
+		}
+	}
+
+	/**
+	 * Render the admin bar button.
+	 *
+	 * @param WP_Admin_Bar $wp_admin_bar The WP AdminBar object.
+	 */
+	public function add_menu_button( WP_Admin_Bar $wp_admin_bar ) {
+		// TODO: update build name when AMP plugin is updated via usual WP ajax call.
```

Comment:
Is this needed since the page will have reloaded?

Reply:
Yes, as clicking the "update now" link on the plugins page to update the plugin wouldn't update the build name in the admin bar.